bootloader/src/pyi_utils.c

```diff
@@ -1074,210 +1074,304 @@ static const char *CC2Str(FourCharCode code) {
     return buf;
 }
 
-/* Handles apple events 'odoc' and 'GURL', both before and after the child_pid is up, forwarding them to args if child
- * not up yet, or otherwise forwarding them to the child if the child is started. Other event types are ignored. */
-static pascal OSErr handle_apple_event(const AppleEvent *theAppleEvent, AppleEvent *reply, SRefCon handlerRefCon)
+/* Generic event forwarder -- forwards an event to the child process. */
+static OSErr generic_forward_apple_event(const AppleEvent *const theAppleEvent, const AEEventClass eventClass,
+                                         const AEEventID evtID, const char *const descStr)
 {
-    const FourCharCode evtCode = (FourCharCode)handlerRefCon;
-    /* Note: the single-quoted 'odoc' & 'GURL' below are not a typo. They are the way
-     * the Apple API encodes UInt32 in a "programmer-friendly" manner using the ISO-C
-     * multi-character constant language feature. These evaluate to UInt32, which is
-     * what the "FourCharCode" type is typedef'd as. */
-    const Boolean apple_event_is_open_doc = evtCode == 'odoc',
-                  apple_event_is_open_uri = evtCode == 'GURL';
+    const FourCharCode evtCode = (FourCharCode)evtID;
+    OSErr err;
+    AppleEvent childEvent;
+    AEAddressDesc target;
+    DescType actualType = 0, typeCode = typeWildCard;
+    char *buf = NULL; /* dynamic buffer to hold copied event param data */
+    Size bufSize = 0, actualSize = 0;
+
+    VS("LOADER [AppleEvent]: Forwarder called for \"%s\".\n", descStr);
+    if (!child_pid) {
+        /* Child not up yet -- there is no way to "forward" this before child started!. */
+         VS("LOADER [AppleEvent]: Child not up yet (child_pid is 0)\n");
+         return errAEEventNotHandled;
+    }
+    VS("LOADER [AppleEvent]: Forwarding '%s' event.\n", CC2Str(evtCode));
+    err = AECreateDesc(typeKernelProcessID, &child_pid, sizeof(child_pid), &target);
+    if (err != noErr) {
+        OTHERERROR("LOADER [AppleEvent]: Failed to create AEAddressDesc: %d\n", (int)err);
+        goto out;
+    }
+    VS("LOADER [AppleEvent]: Created AEAddressDesc.\n");
+    err = AECreateAppleEvent(eventClass, evtID, &target, kAutoGenerateReturnID, kAnyTransactionID,
+                             &childEvent);
+    if (err != noErr) {
+        OTHERERROR("LOADER [AppleEvent]: Failed to create event copy: %d\n", (int)err);
+        goto release_desc;
+    }
+    VS("LOADER [AppleEvent]: Created AppleEvent instance for child process.\n");
 
-    if (apple_event_is_open_doc || apple_event_is_open_uri) {
-        const char *const descStr = apple_event_is_open_uri ? "GetURL" : "OpenDoc";
 
-        VS("LOADER [AppleEvent]: %s handler called.\n", descStr);
+    if (!theAppleEvent) {
+        VS("LOADER [AppleEvent]: New AppleEvent class: '%s' code: '%s'\n",
```

Comment:
Perhaps a comment on why the event-to-be-forwarded could ever be NULL here would be nice. Perhaps also at the top of the function, it could be mentioned that theAppleEvent being NULL is used internally to create and dispatch a new event?

Reply:
Yeah ok sure -- good idea.

Reply:
Ok, done.